sandbox/cgroup/tracking.go

```diff
@@ -0,0 +1,282 @@
+		return ErrCannotTrackProcess
+	}
+	return nil
+}
+
+// ConfirmSystemdServiceTracking checks if systemd tracks this process as a snap service.
+//
+// Systemd is placing started services, both user and system, into appropriate
+// tracking groups. Given a security tag we can confirm if the current process
+// belongs to such tracking group and thus could be identified by snapd as
+// belonging to a particular snap and application.
+//
+// If the application process is not tracked then ErrCannotTrackProcess is returned.
+func ConfirmSystemdServiceTracking(securityTag string) error {
+	pid := osGetpid()
+	path, err := cgroupProcessPathInTrackingCgroup(pid)
+	if err != nil {
+		return err
+	}
+	unitName := fmt.Sprintf("%s.service", securityTag)
+	if !strings.Contains(path, unitName) {
+		return ErrCannotTrackProcess
+	}
+	return nil
+}
+
+func sessionOrMaybeSystemBus(uid int) (isSessionBus bool, conn *dbus.Conn, err error) {
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus. We have a preference for session bus, as this is
+	// where applications normally go to. When a session bus is not available
+	// and the invoking user is root, we use the system bus instead.
+	//
+	// It is worth noting that hooks will not normally have a session bus to
+	// connect to, as they are invoked as descendants of snapd, and snapd is a
+	// service running outside of any session.
+	conn, err = dbusutil.SessionBus()
+	if err == nil {
+		logger.Debugf("using session bus")
+		return true, conn, nil
+	}
+	logger.Debugf("session bus is not available: %s", err)
+	if uid == 0 {
+		logger.Debugf("falling back to system bus")
+		conn, err = dbusutil.SystemBus()
+		if err != nil {
+			logger.Debugf("system bus is not available: %s", err)
+		} else {
+			logger.Debugf("using system bus now, session bus was not available")
+		}
+	}
+	return false, conn, err
+}
+
+type handledDBusError struct {
+	msg       string
+	dbusError string
+}
+
+func (e *handledDBusError) Error() string {
+	return fmt.Sprintf("%s [%s]", e.msg, e.dbusError)
+}
+
+var (
+	errDBusUnknownMethod    = &handledDBusError{msg: "unknown dbus object method", dbusError: "org.freedesktop.DBus.Error.UnknownMethod"}
+	errDBusNameHasNoOwner   = &handledDBusError{msg: "dbus name has no owner", dbusError: "org.freedesktop.DBus.Error.NameHasNoOwner"}
+	errDBusSpawnChildExited = &handledDBusError{msg: "dbus spawned child process exited", dbusError: "org.freedesktop.DBus.Error.Spawn.ChildExited"}
+)
+
+// doCreateTransientScope creates a systemd transient scope with specified properties.
+//
+// The scope is created by asking systemd via the specified DBus connection.
+// The unit name and the PID to attach are provided as well. The DBus method
+// call is performed outside confinement established by snap-confine.
+var doCreateTransientScope = func(conn *dbus.Conn, unitName string, pid int) error {
+	// The property and auxUnit types are not well documented but can be traced
```

Comment:
If there's any reference to code in systemd source tree, it'd be worth to add it here.

Reply:
Done